### PR TITLE
Merge the implementation of GraphHistory and GraphHistoryMultiStudies

### PR DESCRIPTION
**Pull request by keisuke-umezawa**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

https://github.com/optuna/optuna-dashboard/issues/466

### Code to run this feature
In this code, Study0 uses user_attrs of "x1" and "x2" and Study1 uses user_attrs of "x1" and "x". It is expected to show only the intersection of them, and it means that it will show "x1".
```
import logging

import optuna
from optuna_dashboard._app import create_app


host = "127.0.0.1"
port = 8080

optuna.logging.set_verbosity(logging.CRITICAL)


def create_optuna_storage() -> optuna.storages.InMemoryStorage:
    storage = optuna.storages.InMemoryStorage()

    # Single-objective study with attrs
    study = optuna.create_study(study_name="single-objective-with-attrs", storage=storage)

    def objective_single(trial: optuna.Trial) -> float:
        x1 = trial.suggest_float("x1", 0, 10)
        x2 = trial.suggest_float("x2", 0, 10)
        x3 = trial.suggest_categorical("x3", ["foo", "bar"])
        trial.set_user_attr("x1", x1)
        trial.set_user_attr("x2", x2)
        return (x1 - 2) ** 2 + (x2 - 5) ** 2

    study.optimize(objective_single, n_trials=100)

    # Single objective study with 'inf', '-inf', or 'nan' value with attrs
    study = optuna.create_study(study_name="single-inf-with-attrs", storage=storage)

    def objective_single_inf(trial: optuna.Trial) -> float:
        x = trial.suggest_float("x", -10, 10)
        trial.set_user_attr("x1", x)
        trial.set_user_attr("x", x)
        if trial.number % 3 == 0:
            return float("inf")
        elif trial.number % 3 == 1:
            return float("-inf")
        else:
            return x**2

    study.optimize(objective_single_inf, n_trials=50)

    return storage


def main() -> None:
    storage = create_optuna_storage()
    app = create_app(storage, debug=True)
    app.run(host=host, port=port, reloader=True)


if __name__ == "__main__":
    main()
```

### Screenshot
The screen is as follows and is correct:
スクリーンショット 2023-07-09 15 45 34

### CODE REVIEW

File: optuna_dashboard/ts/trialFilter.ts [hunk 1]
```diff
@@ -192,3 +192,58 @@ export const useObjectiveAndUserAttrTargets = (
...
+  )
+  const minDirections = useMemo<number>(() => {
+    return studies.reduce((acc, study) => {
+      return Math.min(acc, study.directions.length)
+    }, Number.MAX_VALUE)
+  }, [studies])
+
+  const intersect = (arrays: AttributeSpec[][]) => {
+    const deepEqual = (obj1: AttributeSpec, obj2: AttributeSpec) => {
+      return JSON.stringify(obj1) === JSON.stringify(obj2)
```
Comment (c-bata) [inline optuna_dashboard/ts/trialFilter.ts:211]: Screenshot 2023-07-28 18 00 25

How about directly checking the `obj1.key === obj2.key`?

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request. I left some suggestions.

File: optuna_dashboard/ts/trialFilter.ts (hunk 1 above)
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/trialFilter.ts]: Fixed it.

File: optuna_dashboard/ts/components/StudyHistory.tsx [hunk 2]
```diff
@@ -89,19 +89,21 @@ export const StudyHistory: FC<{ studyId: number }> = ({ studyId }) => {
...
+        >
+          <CardContent>
+            <GraphHistory
+              studies={[studyDetail]}
+              includePruned={includePruned}
+              logScale={logScale}
+            />
+          </CardContent>
+        </Card>
+      ) : null}
```
[the remark this answers is not in the document]
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/StudyHistory.tsx]: Fixed it.

**Review by c-bata — APPROVED** [review]
Sorry for the late response. LGTM!

**Pull request merged**